firedrake/deflation.py

```diff
@@ -0,0 +1,226 @@
+            return
+        viewer.printfASCII("Firedrake deflated SNES\n")
+
+        ctx = get_appctx(snes.getDM())
+        appctx = ctx.appctx
+        deflation = appctx.get("deflation")
+        viewer.printfASCII(f"Deflating {len(deflation.roots)} solutions\n")
+
+        self.inner.view(viewer)
+
+    def solve(self, snes, b, x):
+        from firedrake import Function
+        out = self.inner.solve(b, x)
+        snes.reason = self.inner.reason
+
+        # Record the solution we've just found
+        self.deflation.append(Function(self.problem.u))
+
+        return out
+
+
+class DeflatedKSP:
+    """A custom Python class that implements the key formulae for deflation
+    after solving the linear system with the inner KSP."""
+
+    def __init__(self, deflation, y, ksp, snes):
+        self.deflation = deflation
+        self.y = y
+        self.ksp = ksp
+        self.snes = weakref.proxy(snes)
+
+    def solve(self, ksp, b, dy_pet):
+        # Use the inner ksp to solve the original problem
+        self.ksp.setOperators(*ksp.getOperators())
+        self.ksp.solve(b, dy_pet)
+        deflation = self.deflation
+
+        if self.snes.getType().startswith("vi"):
+            vi_inact = self.snes.getVIInactiveSet()
+        else:
+            vi_inact = None
+
+        # Compute the scaling of the Newton update that
+        # is the net effect of deflation. This is the key step.
+        tau = self.compute_tau(deflation, self.y, dy_pet, vi_inact)
+        dy_pet.scale(tau)
+
+        ksp.setConvergedReason(self.ksp.getConvergedReason())
+
+    def compute_tau(self, deflation, state, update_p, vi_inact):
+        if deflation is not None:
+            Edy = self.getEdy(deflation, state, update_p, vi_inact)
+
+            minv = 1.0 / deflation.evaluate(state)
+            tau = 1/(1 - minv*Edy)
+            return tau
+        else:
+            return 1
+
+    def getEdy(self, deflation, y, dy, vi_inact):
+        if len(deflation) == 0:
+            return 0
+
+        with deflation.deriv(y).dat.vec as deriv:
+            if vi_inact is not None:
+                deriv_ = deriv.getSubVector(vi_inact)
+            else:
+                deriv_ = deriv
+
+            out = -deriv_.dot(dy)
+
+            if vi_inact is not None:
+                deriv.restoreSubVector(vi_inact, deriv_)
+
+        return out
+
+    def reset(self, ksp):
+        self.ksp.reset()
+
+    def view(self, ksp, viewer):
+        self.ksp.view(viewer)
+
+
+class Deflation:
+    """
+    The shifted deflation operator presented in doi:10.1137/140984798.
+    Defaults to power 2, shift 1, and the L2 norm for distance.
+    """
+    def __init__(self, roots=None, power=2, shift=1, op=None):
+        self.power = power
+        self.shift = shift
+        self.roots = list(roots) if roots else []
+
+        if op is None:
+            op = lambda x, y: inner(x - y, x - y)*dx
+        self.op = op
+
+        self.append = self.roots.append
```

Comment:
Is this still necessary?

Reply:
It's convenient to have, if not necessary.